File: java/src/main/java/org/mydotey/quantileestimator/ckms/CKMSQuantileEstimator.java
package org.mydotey.quantileestimator.ckms;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import org.mydotey.quantileestimator.QuantileEstimator;
import org.mydotey.quantileestimator.ckms.CKMSQuantiles.Quantile;

/**
 * @author koqizhao
 *
 * Apr 2, 2018
 */
public class CKMSQuantileEstimator<T> implements QuantileEstimator<T> {

    private CKMSQuantileEstimatorConfig<T> _config;
    private CKMSQuantiles<T> _ckmsQuantiles;

    public CKMSQuantileEstimator(CKMSQuantileEstimatorConfig<T> config) {
        Objects.requireNonNull(config, "config is null");

        _config = config;

        List<Quantile> quantiles = new ArrayList<>();
        _config.getQuantileConfigs().forEach(qc -> {
            Quantile quantile = new Quantile(qc.getQuantile(), qc.getError());
            quantiles.add(quantile);
        });
        _ckmsQuantiles = new CKMSQuantiles<>(quantiles, _config.getComparator(), 500);
    }

    @Override
    public void add(T value) {
        _ckmsQuantiles.insert(value);
    }

    @Override
    public Map<Double, T> get(List<Double> quantiles) {
        if (_ckmsQuantiles.isEmpty())
            return null;

        HashMap<Double, T> results = new HashMap<>();
        for (Double quantile : quantiles) {
            T result = _ckmsQuantiles.get(quantile);
            results.put(quantile, result);
        }

        return results;
    }

}

File: dotnet/src/MyDotey.QuantileEstimator/Ckms/CkmsQuantileEstimator.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;

/**
 * @author koqizhao
 *
 * Apr 2, 2018
 */
namespace MyDotey.Quantile.Ckms
{
    public class CkmsQuantileEstimator<T> : IQuantileEstimator<T>
    {
        private CkmsQuantileEstimatorConfig<T> _config;
        private CKMSQuantiles<T> _ckmsQuantiles;

        public CkmsQuantileEstimator(CkmsQuantileEstimatorConfig<T> config)
        {
            if (config == null)
                throw new ArgumentNullException("config is null");

            _config = config;

            List<CKMSQuantiles<T>.Quantile> quantiles = new List<CKMSQuantiles<T>.Quantile>();
            _config.QuantileConfigs.ForEach(qc =>
            {
                CKMSQuantiles<T>.Quantile quantile = new CKMSQuantiles<T>.Quantile(qc.Quantile, qc.Error);
                quantiles.Add(quantile);
            });
            _ckmsQuantiles = new CKMSQuantiles<T>(quantiles, _config.Comparer);
        }

        public virtual void Add(T value)
        {
            _ckmsQuantiles.Insert(value);
        }

        public virtual Dictionary<Double, T> Get(List<Double> quantiles)
        {
            Dictionary<Double, T> results = null;
            foreach (Double quantile in quantiles)
            {
                T result = _ckmsQuantiles.Get(quantile);
                if (_ckmsQuantiles.IsEmpty)
                    return null;

                if (results == null)
                    results = new Dictionary<Double, T>();

                results[quantile] = result;
            }

            if (results == null)
                results = new Dictionary<Double, T>();

            return results;
        }
    }
}